[c10] Move exception to C10

Yangqing opened the pull request:
There are still a few work to be done:

- Temporarily disabled AT_WARN, as we move logging, that can be handled by error logging.
- A few header files are still being plumbed through
- caffe2::EnforceNotMet aliasing is not done yet
- need to unify the macros. See c10/util/Exception.h

smessmer commented on aten/src/ATen/core/Backtrace.h:
```diff
@@ -1,28 +1,3 @@
-#pragma once
+#include "c10/util/Backtrace.h"
```
The reason for leaving such forwarding headers behind is mostly backwards compatibility because legacy third party code we don't have control over might include them. I don't think that applies to `ATen/core` since the folder is so new. We can probably just delete this header and refactor the files including it.

Yangqing replied:
Thanks for pointing it out - yes, this one is mainly because there are too many indirect dependencies caused (mainly stl headers) so I put it here intending to remove it in an immediate follow up PR.

smessmer commented on aten/src/ATen/core/Error.h:
```diff
@@ -1,194 +1 @@
-  if (!(cond)) {                    \
-    AT_ERROR(at::str(__VA_ARGS__)); \
-  }
+#include "c10/util/Exception.h"
```
same here

Yangqing replied:
ack :)

smessmer commented on aten/src/ATen/native/DispatchStub.cpp:
```diff
@@ -20,7 +20,7 @@ static CPUCapability compute_cpu_capability() {
       return CPUCapability::DEFAULT;
     }
-    AT_WARN("ignoring invalid value for ATEN_CPU_CAPABILITY: ", envar);
+    // AT_WARN("ignoring invalid value for ATEN_CPU_CAPABILITY: ", envar);
```
what's the issue here? Why comment it out?

Yangqing replied:
It's mainly because LOG(ERROR) would be a better long term solution - it also emerges in fb infrastructure better as the standard logging mechanism. I removed Warning class for now but LOG(ERROR) is dependent on glogs move, so I will convert them in an immediate followup PR once this passes.

smessmer commented on aten/src/ATen/native/cudnn/RNN.cpp:
```diff
@@ -1197,7 +1197,7 @@ std::pair<Tensor, hidden_type> _cudnn_impl(
       input, params, has_biases, mode, hidden_size, num_layers, bidirectional);
   if (!weight_buf.defined()) {
-    AT_WARN(WEIGHT_FORMAT_WARN);
+    // AT_WARN(WEIGHT_FORMAT_WARN);
```
also here

Yangqing replied:
ack

smessmer commented on aten/src/ATen/native/cudnn/RNN.cpp:
```diff
@@ -1227,7 +1227,7 @@ std::pair<Tensor, hidden_type> _cudnn_impl(
       input, params, has_biases, mode, hidden_size, num_layers, bidirectional);
   if (!weight_buf.defined()) {
-    AT_WARN(WEIGHT_FORMAT_WARN);
+    // AT_WARN(WEIGHT_FORMAT_WARN);
```
and here. If there's a good reason for commenting it out, let's add a TODO with an explanation of why and when the line can be re-enabled. If it's meant to never be re-enabled, we should remove the line.

Yangqing replied:
ack

smessmer commented on c10/util/Backtrace.h:
```diff
@@ -0,0 +1,14 @@
+#include "c10/macros/Macros.h"
+
+namespace c10 {
+C10_API std::string get_backtrace(
```
I thought you recently changed some sites in caffe2 to `CAFFE2_API`. As a general plan, do we want to keep separate macros for c10/c2/aten or just use `C10_API` everywhere?

smessmer commented on c10/util/Exception.h:
```diff
@@ -0,0 +1,127 @@
+
+ public:
+  Error(
+      const std::string& msg,
```
Let's pass `msg` and `backtrace` by value.

smessmer commented on c10/util/Exception.h:
```diff
@@ -0,0 +1,127 @@
+// TODO: merge AT_CHECK with AT_ASSERTM. CHECK in fbcode means strict failure if not
+// met.
+
+#define AT_ERROR(...) \
```
Do we maybe want to rename these to `C10_xxx`?

Yangqing replied:
Yeah, too many callsites, so I had to defer it to a follow up PR :)

smessmer commented on c10/util/StringUtil.cpp:
```diff
@@ -0,0 +1,43 @@
+  size_t numReplaced = 0;
+  std::string::size_type lenFrom = std::strlen(from);
+  std::string::size_type lenTo = std::strlen(to);
+  for (std::string::size_type pos = s.find(from); pos != std::string::npos;
```
nit: loop index variables are a perfect place for increasing readability and reducing error likelihood with `auto`.

Yangqing replied:
Good point. will do.

smessmer commented on torch/csrc/autograd/functions/comm.cpp:
```diff
@@ -86,10 +86,10 @@ variable_list Gather::apply(variable_list&& inputs) {
-        "Was asked to gather along dimension 0, but all "
-        "input tensors were scalars; will instead unsqueeze "
-        "and return a vector.");
+    // AT_WARN(
```
same here

Yangqing replied:
ack

smessmer commented on torch/csrc/autograd/python_anomaly_mode.cpp:
```diff
@@ -34,8 +34,8 @@ void PyAnomalyMetadata::print_stack() {
   if (!stack) {
-    AT_WARN("No forward pass information available. Enable detect anomaly "
-            "during forward pass for more information.");
+    // AT_WARN("No forward pass information available. Enable detect anomaly "
```
here

Yangqing replied:
ack

smessmer commented on torch/csrc/autograd/python_anomaly_mode.cpp:
```diff
@@ -51,8 +51,8 @@ void PyAnomalyMetadata::print_stack() {
 
-  AT_WARN("Traceback of forward call that caused the error:\n",
-          THPUtils_unpackString(msg.get()));
+  // AT_WARN("Traceback of forward call that caused the error:\n",
```
and here

Yangqing replied:
ack

smessmer reviewed (approved):
Some comments are about things not changed by this diff (i.e. nits in the code you moved to another file). Feel free to either fix or ignore them.

Also just noticed that you're actually addressing a few of my comments in your diff description already. Should have read that before ;)

Yangqing commented:
@soumith who is the most knowledgeable person to untangle the many libraries under torch/? In this case libshm is not being able to link against c10...

Yangqing commented:
nvm, tools/build_torch_lib.{sh,bat} is... interesting.

Yangqing committed:
Move exception to C10 (#12354)

Summary:
There are still a few work to be done:

- Move logging and unify AT_WARN with LOG(ERROR).
- A few header files are still being plumbed through, need cleaning.
- caffe2::EnforceNotMet aliasing is not done yet.
- need to unify the macros. See c10/util/Exception.h

This is mainly a codemod and not causing functional changes. If you find your job failing and trace back to this diff, usually it can be fixed by the following approaches:

(1) add //caffe2/c10:c10 to your dependency (or transitive dependency).
(2) change objects such as at::Error, at::Optional to the c10 namespace.
(3) change functions to the c10 namespace. Especially, caffe2::MakeString is not overridden by the unified c10::str function. Nothing else changes.

Please kindly consider not reverting this diff - it involves multiple rounds of rebasing and the fix is usually simple. Contact jiayq@ or AI Platform Dev for details.

Pull Request resolved: https://github.com/pytorch/pytorch/pull/12354

Reviewed By: bwasti

Differential Revision: D10238910

Pulled By: bwasti

fbshipit-source-id: 3a5010046eb7d36c2d8ac825307977b518113161

wanchaol commented on caffe2/utils/Optional.h:
```diff
@@ -179,13 +179,11 @@ constexpr struct in_place_t{} in_place{};
 
```
Hi @Yangqing, one simple question here: why we need to keep this version of optional in caffe2/utils, given that we already have a optional.h in c10?
